### ISSUE TITLE
Using external Python variables in "Algorithm"

### ISSUE THREAD

**Issue by bokveizen** [Felix-Petersen/algovision#2]
Really thank you for your wonderful work and code!
I am wondering how I should use external variables in an "Algorithm" environment. For example, if I have a list whose elements' gradients are not necessary (i.e., treated as constant numbers), and I want to use the elements in an "Algorithm", then what should I do? An illustrative example would be that on top of the current bubble sort, I have another "offset" array, before the sorting process, I would like to add the offset array to the input array, then what should I do? Is it possible to write something like
```python
offset = torch.randn(1, 8, requires_grad=False)
Algorithm(
    Input('array'),
    # variables...
    VarInt('n', lambda array: array.shape[1] - 1),
    Var('temp', torch.tensor(0.)),
    For('i', 'n',
        Let('temp', 'array', ['i']),
        Let('array', ['i'], lambda temp: temp + offset[lambda i:i]),
    ),
    # current bubble sort process...
)
```
If this is illegal, what should I do? Is it necessary to include "offset" as an "Input" too, i.e., something like this:
```python
# offset = torch.randn(1, 8, requires_grad=False)
Algorithm(
    Input('array'),
    Input('offset'),
    # variables...
    VarInt('n', lambda array: array.shape[1] - 1),
    Var('temp1', torch.tensor(0.)),
    Var('temp2', torch.tensor(0.)),
    For('i', 'n',
        Let('temp1', 'array', ['i']),
        Let('temp2', 'offset', ['i']),
        Let('array', ['i'], lambda temp1, temp2: temp1 + temp2),
    ),
    # current bubble sort process...
)
```

**Comment by Felix-Petersen** [comment]
Thank you for opening the issue!

This should work except for one issue: A lambda expression within a lambda expression is not supported. (This cannot be supported due to python internals; only the outer lambda can be inspected by algovision for retrieving the variables.) However, this can easily be fixed by writing.

```python
Let('array', ['i'], lambda temp, i: temp + offset[i]),
```
Just make sure that the shapes of `temp + offset[i]` match to avoid bugs through PyTorch's automatic broadcasting. E.g. via
```python
Let('array', ['i'], lambda temp, i: temp + offset[0, i]),
```
Also, the later example would also be possible and probably better if `offset` was a variable that can also change.

A full working example for your code is
```python
from algovision import *
import torch

torch.manual_seed(0)
offset = torch.randn(1, 8)  # requires_grad is by default False

algo = Algorithm(
    Input('array'),
    # variables...
    VarInt('n', lambda array: array.shape[1] - 1),
    Var('temp', torch.tensor(0.)),
    For('i', 'n',
        Let('temp', 'array', ['i']),
        Let('array', ['i'], lambda temp, i: temp + offset[0, i]),
    ),
    # current bubble sort process...
    Output('array')
)

print(offset)
print(algo(torch.arange(5).view(1, 5)))
```
which returns 
```
tensor([[ 1.5410, -0.2934, -2.1788,  0.5684, -1.0845, -1.3986,  0.4033,  0.8380]])
tensor([[ 1.5410,  0.7066, -0.1788,  3.5684,  4.0000]])
```

**Comment by bokveizen** [comment]
Thanks a lot for your explanation and help!
